src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs

```diff
@@ -86,12 +92,12 @@ public string RedirectUri
             set => SetBool(RefreshKey, value);
         }
 
-        private string GetString(string key)
+        protected string GetString(string key)
```

Comment:
Two more methods to add: `public/protected object GetParameter(string name)` that does the same TryGetValue to avoid dictionary throwing. And add SetParemeter for symmetry.

Reply:
Generic overload that attempts a type cast yes/no?

Reply:
I think we should see what the actual derived ChallengeProperties look like before adding the sugar/helpers that we need for them...  But I don't disagree likely we'll need more helpers for the parameters here...

I just think we should start with none, and add them once we need them

Reply:
Basically I don't think we'll be able to properly review/design the base class helpers without seeing what the ChallengeProperties actually look like for real

Reply:
We don't want to do too many trips back and forth, x-repo development is even more of a pain for external users. A single GetParameter method is better than nothing. Don't worry about the generic.

Reply:
I'm not sure its any less painful for us :)  But I'd actually recommend doing both PRs in parallel instead of back and forth per say...

Start with the initial one to get the type to show up in the official bits (if needed), and then have a parallel one with additional helpers added as you go that gets checked in at the same time as security.

Reply:
But given how this is purely additive, he can probably just create both PRs at the same time and we can merge them together at once they are fully complete

Reply:
We have at least an `ICollection<string>` for the scopes, a `TimeSpan?` for the OIDC’s `MaxAge`, and a `bool` for Google’s `IncludeGrantedScopesParameterKey`.

Having a generic `T GetParameter<T>(string key)` and a `void SetParameter<T>(string key, T value)` would easily allow us to cover all of this, so we wouldn’t need more helpers than those for the properties.

Reply:
Yeah that's how I'd have done it, sounds good to me, do what feels best/cleanest :)

Reply:
The generic only hides a cast. It's not giving you serialization like the other helpers here.

`GetParameter<ICollection<string>>` vs `(ICollection<string>)GetParameter` is pretty much the same.

Reply:
Going back to my point...just implement the actual properties classes...before trying to design the helpers :)

Reply:
I was actually doing the following in my implementation:

```
Parameters.TryGetValue(key, out var obj) && obj is T value ? value : default
```

I would personally rather ignore parameters that have the wrong type than have them raise an exception for them (or to add error handling into the handlers for everything).

Reply:
My personal guess is that we would have a common set of helpers, that takes the Dictionary as an argument.

`GetString(Items | Parameters, key),`

Reply:
The dictionaries are different types...